src/router_core/connections.c

```diff
@@ -305,12 +307,14 @@ int qdr_connection_process(qdr_connection_t *conn)
 
     int event_count = 0;
 
+    sys_mutex_lock(conn->work_lock);
+
     if (conn->closed) {
         conn->protocol_adaptor->conn_close_handler(conn->protocol_adaptor->user_context, conn, conn->error);
+        sys_mutex_unlock(conn->work_lock);
```

Comment:
Can the work lock be unlocked before calling into the conn_close_handler?  The handler's pluggable so we don't control which locks it takes (or may take in the future).

Reply:
Probably. From the TSan warning, the important part is that access to `conn->closed` be protected.

Reply:
done